### ISSUE TITLE
extract a representative set of candidates from the Pareto set in multiobjective optimization

### ISSUE THREAD

**Issue by teytaud** [facebookresearch/nevergrad#498]
We have multiobjective optimization, which produces a possibly huge Pareto set.
We might want to extract a small sample, representative, as far as possible, of the whole.

Possibilities:
- random sampling
- k-means

There is a wide range of applications, in particular for @brozi .

@brockho I'd love to have you with us on these works.

**Comment by brockho** [comment]
I agree with the suggestion and that it is important in practice. The problem is also known as subset selection which might help to find additional (and more recent papers, see below). 

The two above ideas seem to be already good starting points that also many people use according to my limited knowledge of the (E)MO literature, especially the first one. Personally, I would rather try to use a more representative and less random set which is linked to a performance measure of interest and the optimal set of the corresponding subset selection problem. 

In the case of (only) two objective functions and the epsilon and hypervolume indicators as performance measures, I suggest to have a look at the work of Karl Bringmann, Tobias Friedrich, and Patrick Klitzke in [this paper](https://dl.acm.org/doi/pdf/10.1145/2576768.2598276?casa_token=zfTVvJzcLbsAAAAA:n6h1w0GBBQIow5VuOM-aT2JqWiNK9RLE-qPGU8GhH0m6z-mTWFruzlgI6MXejJl1f0eoFg33VYU). It looks like the code is still available online in Java (reference [1] in the paper). 

For more papers related to the topic of (hypervolume based) subset selection, please have a look at the references in the [list of open problems](https://simco.gforge.inria.fr/doku.php?id=openproblems#hypervolume-based_subset_selection_problem) from the SIMCO Lorentz Center workshop. Michael Emmerich and I try to keep this updated, but I see that the last change was made in 2015 for this topic :-(

**Comment by brockho** [comment]
I just added two recent papers to the above referenced [list of open problems](https://simco.gforge.inria.fr/doku.php?id=openproblems#hypervolume-based_subset_selection_problem) of which one proposes efficient algorithms for the 3-objective case if the sought after set is small [GM2019] (unfortunately, I don't have access to the pdf right now to check more details).

**Comment by teytaud** [comment]
@brockho in case you have an idea, I try to implement a first fast thing as a baseline,  is it a good first workaround to implement something like "find N elements in the Pareto set of cardinal M >> N so that their hypervolume is maximum " ?

**Comment by brockho** [comment]
Note to ourselves: in today's discussion, we said that a first simple step is to take the set of points that is the best wrt. hypervolume out of maybe 1000 randomly chosen subsets of N points out of the Pareto set approximation of size M.

Second point to investigate: how large are the actual sets after the optimization (in the GAN example)?

**Comment by SiddharthSingh9018** [comment]
Hi, I was looking into this issue and noticed that `pareto_front()` already supports a `subset` parameter with several selection methods, including `"hypervolume"`, `"loss-covering"`, `"domain-covering"`, and `"random"` (see the current docs/optimization.md example). It looks like this was implemented via the `HypervolumePareto` class at some point after this issue was opened, but the issue was never linked/closed.

Could this be closed, or is there a specific aspect of subset selection still considered open (e.g. improving on the current hypervolume-subset method, or benchmarking it against the papers discussed above)? Happy to help with either the cleanup or a follow-up improvement if there's still work here.
